apps/idos-enclave/src/app.tsx

```diff
@@ -0,0 +1,224 @@
+      event.preventDefault();
+      onError("closed");
+    },
+    [onError],
+  );
+
+  const messageReceiver = useCallback(
+    (event: MessageEvent<EventData>) => {
+      if (event.source !== enclave) return;
+
+      const { data: requestData, ports } = event;
+
+      if (!allowedIntents.includes(requestData.intent))
+        throw new Error(`Unexpected request from parent: ${requestData.intent}`);
+
+      responsePort.current = ports[0];
+      encryptionPublicKey.value = requestData.message?.expectedUserEncryptionPublicKey;
+
+      switch (requestData.intent) {
+        case "auth":
+          method.value = null;
+          break;
+
+        case "password":
+          method.value = "password";
+          break;
+
+        case "confirm":
+          confirm.value = true;
+          origin.value = requestData.message?.origin;
+          message.value = requestData.message?.message;
+          break;
+
+        case "backupPasswordOrSecret":
+          isBackupMode.value = true;
+          backupStatus.value = requestData.message?.status;
+          break;
+      }
+
+      if (requestData.configuration.mode) mode.value = requestData.configuration.mode;
+      if (requestData.configuration.theme) theme.value = requestData.configuration.theme;
+    },
+    [
+      enclave,
+      isBackupMode,
+      backupStatus,
+      confirm,
+      message,
+      origin,
+      theme,
+      mode,
+      encryptionPublicKey,
+      method,
+    ],
+  );
+
+  effect(() => {
+    window.addEventListener("message", messageReceiver);
+    window.addEventListener("beforeunload", onBeforeUnload);
+
+    window.dispatchEvent(new Event("idOS-Enclave:ready"));
+
+    return () => {
+      window.removeEventListener("message", messageReceiver);
+      window.removeEventListener("beforeunload", onBeforeUnload);
+    };
+  });
+
+  const methodProps = {
+    store,
+    onError,
+    onSuccess,
+    mode: mode.value,
+  };
+
+  if (confirm.value && message.value) {
+    return (
+      <Layout>
+        <Confirmation message={message.value} origin={origin.value} onSuccess={onSuccess} />
+      </Layout>
+    );
+  }
+
+  if (method.value === "password") {
+    return (
+      <Layout>
+        <PasswordForm
+          {...methodProps}
+          encryptionPublicKey={encryptionPublicKey.value}
+          userId={userId.value}
+        />
+      </Layout>
+    );
+  }
+
+  if (isBackupMode.value) {
+    return (
+      <Layout>
+        <PasswordOrKeyBackup
+          store={store}
+          onSuccess={onSuccess}
+          backupStatus={backupStatus.value}
+        />
+      </Layout>
+    );
+  }
+
+  if (isRecoveryMode.value) {
+    return (
+      <Layout>
+        <PasswordOrKeyRecovery onSuccess={onSuccess} />
+      </Layout>
+    );
+  }
+
+  return (
+    <Layout>
+      <PasswordForm
+        {...methodProps}
+        encryptionPublicKey={encryptionPublicKey.value}
+        userId={userId.value}
+      />
+    </Layout>
+  );
```

Comment:
Don't action: this, in my eyes, begs for the same type of union style we ended up at with idOSClient.